tools/schema/schema.py

```diff
@@ -62,6 +62,9 @@ def generate_project_schema() -> str:
         "$ref": "#/definitions/Part"
     }
     project_schema["definitions"]["Part"] = part_schema
+    # allow "plugin" under "parts" to have extra fields such as "python-packages"
```

Comment:
isn't this overly permissive now? It means that typos like `stage-packagse` won't be caught be the schema anymore. That sounds worse than getting validation errors for the (comparatively few) plugin-specific parameters.
There is a way, on the Rockcraft-side, to crawl the available plugins and see each of their possible properties. Is there a way to express this on the schema? Like, if `plugin` is `python` then the set of valid properties for the part is extended with `python-requirements`, etc

Reply:
> isn't this overly permissive now? It means that typos like `stage-packagse` won't be caught be the schema anymore. That sounds worse than getting validation errors for the (comparatively few) plugin-specific parameters. There is a way, on the Rockcraft-side, to crawl the available plugins and see each of their possible properties. Is there a way to express this on the schema? Like, if `plugin` is `python` then the set of valid properties for the part is extended with `python-requirements`, etc

We can probably enumerate all plugin properties and make our schema whitelist them only, but there doesn't seem a way to limit each a subset of those plugin properties to one plugin value. In that case, the schema would consider `python-packages' field valid for `plugin: npm`.

Reply:
Actually, it turns out we can. After diving deeper into jsonschema docs, I found that it is possible to achieve by using [if-then-else](https://json-schema.org/understanding-json-schema/reference/conditionals#ifthenelse).

Reply:
> That sounds worse than getting validation errors for the (comparatively few) plugin-specific parameters.

I'd say the opposite. There are 2 possible scenarios:
 1. `additionalProperties: false` - users get "False Positives" in their schema validation, misleading them to think their project is wrong and thus spend time trying to fix it, when in fact everything is ok.
 2. `additionalProperties: true` - users get no feedback, a false sense of security, and will eventually build a project that might contain errors. **However**, Rockcraft does project validation within, so whether it is through the JSON schema or the Rockcraft Pydantic validation, the user will always get feedback. 

Given that we are introducing the schema validation, the most logical is to be conservative (as always), and tackle it iteratively - i.e. _validate what we can_. Option 1. is dangerous IMO, as it provides a wrong message, while 2. is simply, inaccurate.

Having said this, if there's a way to add those fields to the schema, I'd obviously prefer that :)

Reply:
I think we should iterate on adding those fields to the schema, but I'd like to chip in that in my opinion not all `part` properties are created equal: many, possibly most parts use either the `nil` or `dump` plugins which have no extra properties. So for those cases allowing additional properties has only downsides and no false positives

Reply:
I added plugin property names to the scheam using schema conditionals in the latest commits.